### PR TITLE
TFT model: separate layers implementation

### PR DESCRIPTION
**Pull request by egoriyaa**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: etna/models/nn/tft_native/layers.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        self.dropout = dropout
+        self.context = context
+
+        self.fc1 = nn.Linear(self.input_size, self.input_size)
+        self.elu = nn.ELU()
+
+        if self.context:
+            self.context_fc = nn.Linear(self.input_size, self.input_size, bias=False)
+
+        self.residual_fc = nn.Linear(self.input_size, self.output_size) if self.input_size != self.output_size else None
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/layers.py:128]: What is it?
In paper notation all the linear layers shouldn't change the shape, the same in GLU input_size=output_size
Снимок экрана 2024-02-09 в 09 04 27

Comment (alex-hse-repository) [reply etna/models/nn/tft_native/layers.py:128]: There are also words about "weights sharing"
Снимок экрана 2024-02-09 в 09 06 47

Doesn't it imply that all the layers should share weights?

Comment (alex-hse-repository) [reply etna/models/nn/tft_native/layers.py:128]: UPD: it means weights sharing across time dim

File: etna/models/nn/tft_native/layers.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        -------
+        :
+            output batch of data with shapes (batch_size, num_timestamps, output_size)
+        """
+        x = self.glu(x)
+        x = self.norm(x + residual)
+        return x
+
+
+class GatedResidualNetwork(nn.Module):
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/layers.py:99]: Actually the reference implementation here should be implementation from google [see](https://github.com/google-research/google-research/blob/master/tft/libs/tft_model.py)

File: etna/models/nn/tft_native/layers.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+            whether to pass context vector through the block
+        dropout:
+            dropout rate
+        """
+        super().__init__()
+        self.input_size = input_size
+        self.features = features
+        self.context = context
+        self.dropout = dropout
+        self.grns = {
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/layers.py:183]: It should be [`ModuleDict`](https://pytorch.org/docs/stable/generated/torch.nn.ModuleDict.html) -- otherwise modules won't be properly registered

File: etna/models/nn/tft_native/layers.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        context:
+            batch of data passed as the context through the block with shapes (batch_size, num_timestamps, num_features * input_size)
+        Returns
+        -------
+        :
+            output batch of data with shapes (batch_size, num_timestamps, output_size)
+        """
+        output = torch.zeros(
+            list(x.values())[0].size() + torch.Size([len(x)])
+        )  # (batch_size, num_timestamps, input_size, num_features)
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/layers.py:228]: Good idea to add comments with shapes 👍

File: etna/models/nn/tft_native/layers.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        for i, (feature, embedding) in enumerate(x.items()):
+            output[:, :, :, i] = self.grns[feature](embedding)
+        flatten_input = torch.cat(list(x.values()), dim=-1)  # (batch_size, num_timestamps, input_size * num_features)
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/layers.py:231]: Better to separate logical blocks with empty lines

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]
Looks good 👍 
1. There might be troubles with shapes, but i guess it will be easier to fix in the next stages
2. What about Interpretable Multi-Head Attention module?

File: etna/models/nn/tft_native/layers.py [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+            output batch of data with shapes (batch_size, num_timestamps, output_size)
+        """
+        x = self.dropout(x)
+        a = self.activation_fc(x)
+        b = self.sigmoid(self.gated_fc(x))
+        x = torch.mul(a, b)
+        return x
+
+
+class GateAddNorm(nn.Module):
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:56]: Actually, yes
It is used in 3 places

File: etna/models/nn/tft_native/layers.py (hunk 1 above)
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:128]: image

We need reshape of residual data in Variable Selection Block, because in the beginning it will be `num_features * hidden_size`, but in the end it must be `num_features`

**Comment by egoriyaa** [comment]
About IMHA
Why i think we shouldn't add it now:
1) If something goes wrong during testing it will be hard to find mistakes (it can be in implementation of IMHA because it is complex or outside). If all is good without IMHA we will change nn.MultiHeadAttention for our implementation
2) Don't add IMHA at all. It plays role only in "interpretation", but not in performance. And I don't think somebody will "Interpret" smth.

File: etna/models/nn/tft_native/layers.py (hunk 2 above)
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:99]: What it means?

Comment (alex-hse-repository) [reply etna/models/nn/tft_native/layers.py:99]: It means that if you use some implementation as reference it is better to be this one

**Comment by alex-hse-repository** [comment]
> And I don't think somebody will "Interpret" smth.

It's not actually true, we have the separate track about NNs interpretation

File: etna/models/nn/tft_native/layers.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+            batch of data passed as the context through the block with shapes (batch_size, num_timestamps, output_size)
+        Returns
+        -------
+        :
+            output batch of data with shapes (batch_size, decoder_length, output_size)
+        """
+        x = self.grn1(x, context)
+        residual = x
+
+        attn_mask = torch.triu(torch.ones(x.size()[1], x.size()[1], dtype=torch.bool), diagonal=1)
```
Comment (d-a-bunin) [inline etna/models/nn/tft_native/layers.py:353]: Should we clarify what is happening here? As I understand, we create a causal attention mask.

File: etna/models/nn/tft_native/layers.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+        attn_mask[:, : -self.decoder_length] = False
```
Comment (d-a-bunin) [inline etna/models/nn/tft_native/layers.py:354]: What it is for?

File: etna/models/nn/tft_native/layers.py [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+
+        Returns
+        -------
+        :
+            output batch of data with shapes (batch_size, num_timestamps, output_size)
+        """
+        x = self.dropout(x)
+        a = self.activation_fc(x)
+        b = self.sigmoid(self.gated_fc(x))
+        x = torch.mul(a, b)
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:52]: `mul` is faster for huge shapes
but may be in our case there is no difference

File: etna/models/nn/tft_native/layers.py [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+    """Gated Residual Network (GRN)."""
+
+    def __init__(self, input_size: int, output_size: int, dropout: float = 0.1, context: bool = False):
+        """Init GRN.
+
+        Parameters
+        ----------
+        input_size:
+            input size of the feature representation
+        output_size:
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:109]: Yes
https://github.com/etna-team/etna/pull/234#discussion_r1484110191

File: etna/models/nn/tft_native/layers.py [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+        :
+            output batch of data with shapes (batch_size, num_timestamps, output_size)
+        """
+        output = torch.zeros(
+            list(x.values())[0].size() + torch.Size([len(x)])
+        )  # (batch_size, num_timestamps, input_size, num_features)
+        for i, (feature, embedding) in enumerate(x.items()):
+            output[:, :, :, i] = self.grns[feature](embedding)
+
+        flatten_input = torch.cat(list(x.values()), dim=-1)  # (batch_size, num_timestamps, input_size * num_features)
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:234]: yes, it is guaranteed according to the docs

we can do `sorted` in model __init__ before passing it to block

File: etna/models/nn/tft_native/layers.py (hunk 7 above)
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:353]: Yes
How can we clarify it?

File: etna/models/nn/tft_native/layers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/models/nn/tft_native/layers.py:52]: Where have you found it?

File: etna/models/nn/tft_native/layers.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/models/nn/tft_native/layers.py:234]: Ok, for the same `dict` it always return them in the same order. But during inference different dicts could be a little different and have different order. I would recommend to iterate `x.values` in order defined by `self.features`.

File: etna/models/nn/tft_native/layers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:52]: measured by myself

File: etna/models/nn/tft_native/layers.py [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+        flatten_grn_output = self.flatten_grn(
+            x=flatten_input, context=context
+        )  # (batch_size, num_timestamps, num_features)
+        feature_weights = self.softmax(flatten_grn_output).unsqueeze(
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:238]: fixed

File: etna/models/nn/tft_native/layers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/models/nn/tft_native/layers.py:52]: I don't really think it is true. Why would they have a different implementation?

[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/layers.py:52]: Why is it so important how I multiply two tensors?
